File: pytemp/pca.py
import numpy as np

def load(filename):
    with open(filename) as fin:
        lines = [line.strip() for line in fin]
        data = np.array([
            [float(el) for el in line.split(',')]
            for line in lines
        ])
        return data[:, :4], data[:, 4:]
        
def demean(X):
    '''X: ndarray'''
    return X - X.mean(0)

def direction(x):
    '''x: ndarray -> ndarray'''
    mag = np.sqrt((x*x).sum())
    return x / mag

def directional_var(data, x):
    d = direction(x)
    x = data.dot(d)
    return (x*x).sum()

def directional_var_grad(data, x):
    d = direction(x)
    return 2 * data.T.dot(data).dot(d)

def first_pc(data, nrounds=100, step_size=0.1):
    guess = np.random.randn(len(data[0]))
    guess = direction(guess)
    for _ in range(nrounds):
        grad = directional_var_grad(data, guess)
        guess -= step_size * grad
    return direction(guess)

def project(direc, vec):
    '''(direc: ndarray, vec: ndarray) -> <vec, direc> direc'''
    return direc.dot(vec) * direc

# project to complement
def project2comp(direc, vec):
    '''(direc: ndarray, vec: ndarray) -> vec - <vec, direc> direc'''
    return vec - project(direc, vec)

def projectem2comp(direc, data):
    '''projects all data onto orthogonal complement of direc'''
    return np.array([project2comp(direc, v) for v in data])

def pca(data, ncomps):
    components = []
    for _ in range(ncomps):
        pc = first_pc(data, 200, 0.1)
        components.append(pc)
        data = projectem2comp(pc, data)
    return np.array(components)

def transform_vec(vec, components):
    return components.dot(vec)

def transformem(data, components):
    return transform_vec(data.T, components).T


if __name__ == '__main__':

    X, y = load('iris.csv')

    # centre the data
    Xc = demean(X)

    pcs = pca(Xc, 3)
    for i, pc in enumerate(pcs, 1):
        print(f'pc {i}', pc)

    Xt = transformem(Xc, pcs)
    print(Xt[:10])

    tvec = transform_vec(Xc[0], pcs)
    print(Xc[0], '->', tvec)

    # from sklearn.decomposition import PCA

    # pm = PCA(n_components=3)
    # X2D = pm.fit_transform(Xc)
    # print(pm.components_)

File: pkg/pca/pca.go
package main

import (
	"fmt"
	"encoding/csv"
	"os"
	"strconv"
	"math"
	"math/rand"
)

// data vector
type Vector []float64

func (vec Vector) Dot(ovec Vector) float64 {
	if len(vec) != len(ovec) {
		fmt.Println("ERROR: vectors must be of the same dimension")
		os.Exit(1)
	}
	var sum float64
	for i, val := range vec {
		sum += ovec[i] * val
	}
	return sum
}

func NewVector(n int) Vector {
	vec := make(Vector, n)
	for i := 0; i < n; i++ {
		vec[i] = rand.NormFloat64()
	}
	return vec
}

func loadData(filename string) ([]Vector, []int) {
	file, err := os.Open("iris.csv")
	if err != nil {
		panic(err)
	}
	defer file.Close()
	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		panic(err)
	}
	X := make([]Vector, len(records))
	y := make([]int, len(records))
	for i, record := range records {
		n := len(record)
		x := make(Vector, n-1)
		for j, item := range record[:n-1] {
			x[j], err = strconv.ParseFloat(item, 64)
			if err != nil {
				fmt.Println("failed to convert element %d in line %d", j, i)
				continue
			}
		}
		X[i] = x
		val, err := strconv.ParseInt(record[n-1], 0, 0)
		if err != nil {
			fmt.Println("failed to parse label in line %d", i)
		}
		y[i] = int(val)  // int64 -> int
	}
	return X, y
}

func demean(data []Vector) []Vector {
	means := make(Vector, len(data[0]))
	for _, row := range data {
		for j, elem := range row {
			means[j] += elem
		}
	}
	N := len(data)
	for i, val := range means {
		means[i] = val / float64(N)
	}
	centred := make([]Vector, len(data))
	for i, row := range data {
		rowc := make(Vector, len(row))  // centred row
		for j, elem := range row {
			rowc[j] = elem - means[j]
		}
		centred[i] = rowc
	}
	return centred
}

func magnitude(vec Vector) float64 {
	var sum float64
	for _, comp := range vec {
		sum += comp * comp
	}
	return math.Sqrt(sum)
}

func direction(vec Vector) Vector {
	mag := magnitude(vec)
	direc := make(Vector, len(vec))
	for i, comp := range vec {
		direc[i] = comp / mag
	}
	return direc
}

func dirVar(data []Vector, vec Vector) float64 {
	var sum float64
	direc := direction(vec) 
	for _, x := range data {
		z := direc.Dot(x)
		sum += z * z
	}
	return sum
}

func dirVarGrad(data []Vector, vec Vector) Vector {
	direc := direction(vec)
	u := make(Vector, len(data))
	for i, x := range data {
		u[i] = x.Dot(direc)
	}
	z := make(Vector, len(vec))
	for i, _ := range vec {
		var sum float64
		for j, val := range u {
			sum += val * data[j][i]
		}
		z[i] = 2 * sum
	}
	return z
}

func FirstPC(data []Vector, nRounds int, stepSize float64) Vector {
	if len(data) == 0 {
		panic("shithead")
	}
	guess := NewVector(len(data[0]))
	for r := 0; r < nRounds; r++ {
		grad := dirVarGrad(data, guess)
		for i, val := range guess {
			guess[i] = val - stepSize * grad[i]
		}
	}
	return direction(guess)
}

func project(direc Vector, vec Vector) Vector {
	pvec := make(Vector, len(direc))
	mag := direc.Dot(vec)
	for i, val := range direc {
		pvec[i] = mag * val
	}
	return pvec
}

func project2comp(direc Vector, vec Vector) Vector {
	pvec := project(direc, vec)
	for i, val := range pvec {
		pvec[i] = vec[i] - val
	}
	return pvec
}

func projectem2comp(direc Vector, data []Vector) []Vector {
	tdata := make([]Vector, len(data))
	for i, vec := range data {
		tdata[i] = project2comp(direc, vec)
	}
	return tdata
}

func PCA(data []Vector, nComps int) []Vector {
	components := make([]Vector, nComps)
	for i := 0; i < nComps; i++ {
		pc := FirstPC(data, 200, 0.1)
		components[i] = pc
		data = projectem2comp(pc, data)
	}
	return components
}

func transform(comps []Vector, vec Vector) Vector {
	tvec := make(Vector, len(comps))
	for i, comp := range comps {
		tvec[i] = comp.Dot(vec)
	}
	return tvec
}

func transformem(comps []Vector, data []Vector) []Vector {
	tvecs := make([]Vector, len(data))
	for i, vec := range data {
		tvecs[i] = transform(comps, vec)
	}
	return tvecs
}

func main() {

	X, _ := loadData("iris.csv")

	Xcentred := demean(X)

	comps := PCA(Xcentred, 3)
	for i, comp := range comps {
		fmt.Printf("pc %d: %v\n", i + 1, comp)
	}

	tdata := transformem(comps, Xcentred)
	for i, v := range tdata[:10] {
		fmt.Printf("vec %d: %v\n", i, v)
	}

	tvec := transform(comps, Xcentred[0])
	fmt.Printf("%v -> %v\n", Xcentred[0], tvec)
	
}

